### PR TITLE
feat(apidocs): add plugin apidocs

### PR DESCRIPTION
**Pull request by bbbmj**
**What this PR does / why we need it**:

Add plugin apidocs to return swagger.json

**Which issue(s) this PR is related to** *(optional, link to 3rd issue(s))*:

Fixes #

Reference to #

**Special notes for your reviewer**:

/cc @liubog2008 

**Release note**:

```release-note
Add plugin apidocs
```

### CODE REVIEW

**Comment by bbbmj** [comment]
usage can be:

code in `cmd/myproject/main.go`

```
	swaggerOption := swagger.NewOption(map[string]string{
		"/apis/v1/swagger.json": "api.v1.json",
		"/apis/v2/swagger.json": "api.v2.json",
	})

	// Enable plugins.
	cmd.EnablePlugin(metricsOption, loggerOption, reqlogOption, versionOption, swaggerOption)
```

**Comment by bbbmj** [comment]
As command to generate swagger.json `nirvana api` depends on binary `nirvana`, it's not friendly to try to get binary `nirvana` in Dockerfile. 

How about making more docs to make it up?

File: plugins/swagger/swagger.go [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+		return "", "", fmt.Errorf("Please specify the swagger config with apiVersion:swaggerFilePath")
+	}
+
+	set := strings.Split(config, ":")
+
+	if len(set) != 2 {
+		return "", "", fmt.Errorf("Please specify the repository name with apiVersion:swaggerFilePath")
+	}
+
+	path := fmt.Sprintf("/apis/%s/swagger.json", strings.TrimSpace(set[0]))
```
Comment (kdada) [inline plugins/swagger/swagger.go]: Don't assume that users will use `/apis/`.

Comment (bbbmj) [reply plugins/swagger/swagger.go]: updated

**Comment by ddysher** [comment]
How useful is it to just serve swagger.json file?  is it possible that we also serve redoc under path `apis/v1`?

one option is that we update our makefile & dockerfile to use two stage build.  The first stage contains a builder image which includes the `nirvana` binary and outputs `swagger.json`, and the second stage serves the file.  we can provide a new command like `make with-apidoc`?
> As command to generate swagger.json nirvana api depends on binary nirvana, it's not friendly to try to get binary nirvana in Dockerfile.

**Comment by bbbmj** [comment]
> How useful is it to just serve swagger.json file? is it possible that we also serve redoc under path apis/v1?

After discussing with @liubog2008 , we decided to just serve swagger.json file as we don't need many ReDoc services in a cluster.

File: plugins/swagger/swagger.go [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+		return nil
+	}
+}
+
+type config struct {
+	config []string
+}
+
+// SetOptions returns a configurer to set swagger options.
+func SetOptions(option Option) nirvana.Configurer {
```
Comment (kdada) [inline plugins/swagger/swagger.go]: Please refer to https://github.com/caicloud/nirvana/blob/master/plugins/reqlog/reqlog.go.
The plugin does not obey the convention of Nirvana.

Option is a convenient way for generating command flags/ENV. It's not for users to set plugin config.

Comment (kdada) [reply plugins/swagger/swagger.go]: Also you need to provide `Default()` and `Disable()`.

**Comment by kdada** [comment]
After examining this use case,  I think you should provide a resource API to cover it.

For example:
`/apidocs` returns a list of the information of all API docs.
`/apidocs/{apidoc}` returns an API doc.

Then users can specify:
1. `--apidocs-path=/apidocs`
2. `--apidocs-files=v1:./api.v1.json,v2:./api.v2.json`

**Comment by bbbmj** [comment]
With this plugin, user can add following to Makefile

```
# Path to your apis
API_DIR ?= pkg/apis

# Api docs output directory.
API_OUTPUT_DIR := ./apis

.PHONY: apidocs

apidocs:
	@nirvana api $(API_DIR) --serve="" --output=$(API_OUTPUT_DIR)
```

File: plugins/apidocs/apidocs.go [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+type apidocsInstaller struct{}
+
+// Name is the external config name.
+func (i *apidocsInstaller) Name() string {
+	return ExternalConfigName
+}
+
+type config struct {
+	path  string
+	files string
```
Comment (kdada) [inline plugins/apidocs/apidocs.go]: Field `files string` should be `files map[string]string`.

File: plugins/apidocs/apidocs.go [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+}
+
+// Name returns plugin name.
+func (p *Option) Name() string {
+	return ExternalConfigName
+}
+
+// Configure configures nirvana config via current options.
+func (p *Option) Configure(cfg *nirvana.Config) error {
+	cfg.Configure(
```
Comment (kdada) [inline plugins/apidocs/apidocs.go]: Then converts `Option.Files` to `config.files`.

File: plugins/apidocs/apidocs.go [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+func (i *apidocsInstaller) Install(builder service.Builder, cfg *nirvana.Config) error {
+	var err error
+	wrapper(cfg, func(c *config) {
+		path, files := i.handle(c)
+
+		var routers []string
+		for k, v := range files {
+			routers = append(routers, k)
+
+			err = builder.AddDescriptor(definition.Descriptor{
```
Comment (kdada) [inline plugins/apidocs/apidocs.go]: Handle this error.

File: plugins/apidocs/apidocs.go [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+						return ioutil.ReadFile(v)
+					},
+				}},
+			})
+		}
+
+		err = builder.AddDescriptor(definition.Descriptor{
+			Path:     path,
+			Consumes: []string{definition.MIMEAll},
+			Produces: []string{definition.MIMEAll},
```
Comment (kdada) [inline plugins/apidocs/apidocs.go]: Produce `json`.

File: plugins/apidocs/apidocs.go [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+			Definitions: []definition.Definition{{
+				Method:  definition.Get,
+				Results: definition.DataErrorResults(""),
+				Function: func(ctx context.Context) ([]string, error) {
+					fmt.Println(routers)
```
Comment (kdada) [inline plugins/apidocs/apidocs.go]: Remove fmt.

File: plugins/apidocs/apidocs.go [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+		path, files := i.handle(c)
+
+		var routers []string
+		for k, v := range files {
+			routers = append(routers, k)
+
+			err = builder.AddDescriptor(definition.Descriptor{
+				Path:     k,
+				Consumes: []string{definition.MIMEAll},
+				Produces: []string{definition.MIMEAll},
```
Comment (kdada) [inline plugins/apidocs/apidocs.go]: Produce `json`.

File: plugins/apidocs/apidocs.go (hunk 5 above)
Comment (kdada) [inline plugins/apidocs/apidocs.go]: Use `Descriptor.Children` instead of `builder.AddDescriptor`.

**Comment by bbbmj** [comment]
@kdada Fixed, PTAL

**Pull request merged**